perfkitbenchmarker/providers/openstack/os_network.py

```diff
@@ -27,134 +29,143 @@
-
-
-class OpenStackPublicNetwork(object):
-
-    def __init__(self, pool):
-        self.__nclient = utils.NovaClient()
-        # TODO(meteorfox) Is this lock needed? Callers to these methods seem
-        # to be using a lock already. Should we make callers use a lock, or
-        # should this object handle the lock?
-        self.__floating_ip_lock = threading.Lock()
-        self.ip_pool_name = pool
-
-    def allocate(self):
-        with self.__floating_ip_lock:
-            return self.__nclient.floating_ips.create(pool=self.ip_pool_name)
-
-    def release(self, floating_ip):
-        f_id = floating_ip.id
-        if self.__nclient.floating_ips.get(f_id):
-            with self.__floating_ip_lock:
-                if self.__nclient.floating_ips.get(f_id):
-                    self.__nclient.floating_ips.delete(floating_ip)
-                    while self.__nclient.floating_ips.findall(id=f_id):
-                        time.sleep(1)
-
-    def get_or_create(self):
-        with self.__floating_ip_lock:
-            floating_ips = self.__nclient.floating_ips.findall(
-                instance_id=None,
-                pool=self.ip_pool_name)
-        if floating_ips:
-            return floating_ips[0]
-        else:
-            return self.allocate()
-
-    def is_attached(self, floating_ip):
-        return self.__nclient.floating_ips.get(
-            floating_ip.id
-        ).fixed_ip is not None
+    if vm.is_static:
+      return
+
+    if to_port is None:
+      to_port = port
+
+    sec_group_rule = (port, to_port, vm.group_id)
+
+    with self._lock:
+      cmd = utils.OpenStackCLICommand(vm, 'security group rule', 'create',
+                                      vm.group_id)
+      cmd.flags['dst-port'] = '%d:%d' % (port, to_port)
+      if sec_group_rule in self.sec_group_rules_set:
+        return
+      for prot in ('tcp', 'udp',):
+        cmd.flags['proto'] = prot
+        cmd.Issue(suppress_warning=True)
+      self.sec_group_rules_set.add(sec_group_rule)
+
+  def DisallowAllPorts(self):
+    """Closes all ports on the firewall."""
+    pass
+
+
+class OpenStackFloatingIPPool(object):
+
+  _floating_ip_lock = threading.Lock()  # Guards floating IP allocation/release
+
+  def __init__(self, pool_name):
+    self.ip_pool_name = pool_name
+
+  def associate(self, vm):
+    with self._floating_ip_lock:
+      floating_ip = self._get_or_create(vm)
+      cmd = utils.OpenStackCLICommand(vm, 'ip', 'floating', 'add',
+                                      floating_ip['ip'], vm.id)
+      del cmd.flags['format']  # Command does not support json output format
+      _, stderr, _ = cmd.Issue()
+      if stderr:
+        raise errors.Error(stderr)
+      return floating_ip
+
+  def _get_or_create(self, vm):
+    list_cmd = [FLAGS.openstack_nova_path, 'floating-ip-list']
+    stdout, stderr, _ = vm_util.IssueCommand(list_cmd)
+    floating_ip_dict_list = utils.ParseFloatingIPTable(stdout)
+    for flip_dict in floating_ip_dict_list:
```

Comment:
Please rename `flip_dict` to something more clear, like `floating_ip`. `ip_dict` would also work. The first time I read the code I wondered what flipping had to do with IPs.